text/0000-monorepo.md

```diff
@@ -0,0 +1,359 @@
+
+- Notifications
+- Issue templates
+- Pull Request templates
+
+These areas will need to be addressed in order to prevent excessive notifications, and provide reasonable options for issue templates to help with triage. Pull request templates are one item that should be shared across the projects as it will provide a formal structure for documenting proposed changes and reviewing them.
+
+For Notifications, most likely we would have to do a custom solution instead of relying on GitHub notifications. There are a couple strategies we could follow:
+
+- Create issue templates based on project areas. These templates could auto-label issues
+- Create a single issue template similar to [`DefinitelyTyped`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/.github/ISSUE_TEMPLATE.md) that includes notifying owners so they trigger a notification
+  - This issue template explains that not notifying owners means that the issue will be ignored
+  - We could use a bot (like stale bot) to auto-close issues without responses in this format
+  - We could create a bot that infers project owners from issue titles if they write out the package name
+  
+This strategy bleeds into issue templates in that we would most likely want a specific structure around issue titles and contents for things like bugs, feature requests, and support.
+
+Pull request templates would ideally follow our current format where they have a corresponding issue and provide a high-level change log and steps to review.
+
+# Drawbacks
+
+The main disadvantages of using the strategies outlined above is that it forces our model into a monorepo. While this in and of itself is not a negative thing, it does shift a portion of our development time towards investing in tooling to help make working in a monorepo more enjoyable.
+
+Aspects of this include a custom CLI to help with working on a single component, in addition to tooling or strategies for scaling out our core tools so that they are either fast on a large number of files, or only run on files that have chanegd.
+
+In addition, while moving work to one project eliminates any confusion to issue creation, we run into a problem where everyone will get additional notifications that may not be related to their core project. The immediate action after this would be to ignore conversations that one is not actively participating in, but this could lead to issues not getting any response.
+
+There are ways to mitigate this, including issue templates, but ultimately scaling issues and pull requests in a monorepo with multiple teams/collaboratos is still an unknown.
+
+Finally, there is a concern around making everything under the Carbon umbrella public. The main use-case being: what if a team wants to use Carbon to build a component but is unable to make this component open-source? One solution could be the mirror our project to an internal GitHub Enterprise repo, but this could cause fragmentation in the ecosystem.
```

Comment:
another option vs mirror the project would be that the component development environment is ingestible for use external the the carbon monorepo. So an add-on repo would pull in carbon as a dependency and the add-on development environment would share a directory structure from carbon when creating components:

```
new-addon-component/
├── new-addon-component-angular
├── new-addon-component-react
├── new-addon-component-vue
├── new-addon-component-handlebars
└── new-addon-component
```

but then if the add-on was creating a change to an existing component, the add-on's development environment would act the same way as developing within Carbon, except any new versions of an existing component would supersede carbon's verion:

```
add-on repo
├── button (carbons, pulled from node_modules as tho local)
├── tooltip (carbons)
├── overflow-menu (addon has different version)
    ├── overflow-menu-angular (carbons, pulled from node_modules as tho local)
    ├── overflow-menu-react (exists within repo)
    ├── overflow-menu-vue (carbons, pulled from node_modules as tho local)
    ├── overflow-menu-handlebars (carbons, pulled from node_modules as tho local)
    └── overflow-menu (exists within repo)
```

For the above, the add-on team's common overflow menu content (like the demos or the selectors object) would differ from Carbon's thus when running tests in the non-react version those implementations would _possibly_ fail some of the tests which would be local to this add-on.

The concept of using this structure instead of an actual mirror would be the ease in updating to the latest carbon. Also, the local overwrite could happen automatically just by adding a new directory with the same naming structure as carbon - ie: adding a `button` directory would _locally_ defer to using the local button component vs carbon's.